### PR TITLE
Fix issue with publishing content with actions

### PR DESCRIPTION
**Pull request by googletorp**

### CODE REVIEW

**Comment by googletorp** [comment]
For reference. The form in core which allows you to change the moderation state in core looks like this:

```
public function submitForm(array &$form, FormStateInterface $form_state) {
    /** @var \Drupal\Core\Entity\ContentEntityInterface $entity */
    $entity = $form_state->get('entity');
    /** @var \Drupal\Core\Entity\ContentEntityStorageInterface $storage */
    $storage = \Drupal::entityTypeManager()->getStorage($entity->getEntityTypeId());
    $entity = $storage->createRevision($entity, $entity->isDefaultRevision());

    $new_state = $form_state->getValue('new_state');

    $entity->set('moderation_state', $new_state);

    if ($entity instanceof RevisionLogInterface) {
      $entity->setRevisionCreationTime($this->time->getRequestTime());
      $entity->setRevisionLogMessage($form_state->getValue('revision_log'));
      $entity->setRevisionUserId($this->currentUser()->id());
    }
    $entity->save();

    $this->messenger()->addStatus($this->t('The moderation state has been updated.'));

    $new_state = $this->moderationInfo->getWorkflowForEntity($entity)->getTypePlugin()->getState($new_state);
    // The page we're on likely won't be visible if we just set the entity to
    // the default state, as we hide that latest-revision tab if there is no
    // pending revision. Redirect to the canonical URL instead, since that will
    // still exist.
    if ($new_state->isDefaultRevisionState()) {
      $form_state->setRedirectUrl($entity->toUrl('canonical'));
    }
  }
```

This is where I got the code from, if this cause double revisions, then content moderation form in core is broken. I think this approach is the best / most robust

File: src/Plugin/Action/StateChange.php [hunk 1]
```diff
@@ -194,7 +204,14 @@ public function calculateDependencies() {
...
   public function execute(ContentEntityInterface $entity = NULL) {
-    $entity = $this->loadLatestRevision($entity);
+    // Create a new revision if entity is revisionable to fix some edge cases
+    // with performing moderation states which publishes the entity.
+    if ($entity->getEntityType()->isRevisionable()) {
+      $entity = $this->loadLatestRevision($entity);
+      $entity = $this->entityTypeManager->getStorage($entity->getEntityTypeId())
+       ->createRevision($entity, $entity->isDefaultRevision());
+    }
+
```
Comment (shaundrong) [inline src/Plugin/Action/StateChange.php]: yep as I said in the meeting, my concern here is duplicate revisions for other entities that use this module. If you don't find any issues with them. I'm fine with this fix.

Comment (shaundrong) [reply src/Plugin/Action/StateChange.php]: you may want to also set the meta data for the revision if you are creating a new revision here. The default createRevision appears to clones those values.

from Drupal\content_moderation\Form\EntityModerationForm
```
    if ($entity instanceof RevisionLogInterface) {
      $entity
        ->setRevisionCreationTime($this->time
        ->getRequestTime());
      $entity
        ->setRevisionUserId($this
        ->currentUser()
        ->id());
    }
```

**Review by shaundrong — CHANGES_REQUESTED** [review]
Please update code post new revision to set metadata about revision per comment.

**Review by shaundrong — APPROVED** [review]
👍

**Pull request merged**